src/bare_index.rs

```diff
@@ -0,0 +1,406 @@
+        }
+    }
+
+    /// Creates and index for the default crates.io registry
+    pub fn crates_io() -> Result<Self, Error> {
+        Self::from_url(crate::INDEX_GIT_URL)
+    }
+
+    /// Opens the local index, which acts as a kind of lock for source control
+    /// operations
+    pub fn open_or_clone(&self) -> Result<BareIndexRepo<'_>, Error> {
+        BareIndexRepo::new(self)
+    }
+}
+
+pub struct BareIndexRepo<'a> {
+    inner: &'a BareIndex,
+    head: git2::Oid,
+    repo: git2::Repository,
+    tree: Option<git2::Tree<'static>>,
+    head_str: String,
+}
+
+impl<'a> BareIndexRepo<'a> {
+    fn new(index: &'a BareIndex) -> Result<Self, Error> {
+        let exists = git2::Repository::discover(&index.path)
+            .map(|repository| {
+                repository
+                    .find_remote("origin")
+                    .ok()
+                    // Cargo creates a checkout without an origin set,
+                    // so default to true in case of missing origin
+                    .map_or(true, |remote| {
+                        remote.url().map_or(true, |url| url == index.url)
+                    })
+            })
+            .unwrap_or(false);
+
+        if !exists {
+            git2::build::RepoBuilder::new()
+                .fetch_options(crate::fetch_opts())
+                .bare(true)
+                .clone(&index.url, &index.path)?;
+        }
+
+        let repo = git2::Repository::open(&index.path)?;
+        let head = repo
+            .refname_to_id("FETCH_HEAD")
+            .or_else(|_| repo.refname_to_id("HEAD"))?;
+        let head_str = head.to_string();
+
+        let tree = {
+            let commit = repo.find_commit(head)?;
+            let tree = commit.tree()?;
+
+            unsafe { std::mem::transmute::<git2::Tree<'_>, git2::Tree<'static>>(tree) }
+        };
+
+        Ok(Self {
+            inner: index,
+            head,
+            head_str,
+            repo,
+            tree: Some(tree),
+        })
+    }
+
+    /// Fetches latest from the remote index repository. Note that using this
+    /// method will mean no cache entries will be used, if a new commit is fetched
+    /// from the repository, as their commit version will no longer match.
+    pub fn fetch(&mut self) -> Result<(), Error> {
+        {
+            let mut origin_remote = self
+                .repo
+                .find_remote("origin")
+                .or_else(|_| self.repo.remote_anonymous(&self.inner.url))?;
+
+            origin_remote.fetch(&["master"], Some(&mut crate::fetch_opts()), None)?;
+        }
+
+        let head = self
+            .repo
+            .refname_to_id("FETCH_HEAD")
+            .or_else(|_| self.repo.refname_to_id("HEAD"))?;
+        let head_str = head.to_string();
+
+        let commit = self.repo.find_commit(head)?;
+        let tree = commit.tree()?;
+
+        let tree = unsafe { std::mem::transmute::<git2::Tree<'_>, git2::Tree<'static>>(tree) };
```

Comment:
Could you make it clearer why it's safe? 

Maybe put both repo and tree in the same struct, so that it's clearer repo can't be touched without breaking tree, or maybe use owning-ref crate.

Reply:
I've added a comment for why this is safe, as the reason I was ok with this particular implementation was because this is essentially the same as how cargo does it. :wink: